### PR TITLE
Batch DataPlane TypeSpec

### LINKED ISSUE #23219: [Azure Batch] API Review
New API Review meeting has been requested.

**Service Name**: Azure Batch
**Review Created By**: Mario Guerra
**Review Date**: 04/20/2023 04:00 PM PT
**PR**: 
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: Azure Batch allows you to run large-scale parallel and batch compute jobs in the Azure cloud. The service is designed to help you easily manage and run large-scale parallel and HPC workloads, without the need for managing the underlying infrastructure. 
The service is GA and has track 1 SDKs that are also GA. The Batch team is moving forward with releasing track 2 SDKs based off TypeSpec. The API is not changing between track 1 and 2. 

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=692d51db-b5c8-ed11-b597-000d3a3418aa)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by wiboris**
# Data Plane API - Pull Request

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document:
* Previous Open API Doc:
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

### CODE REVIEW

**Review by mikekistler — CHANGES_REQUESTED** [review]
There are a few changes needed in this PR.
- please make this a new preview API version, so that we can iterate on it as needed
- I think you need to fix the breaking changes error
> The new version does not support 'application/son; odata=minimalmetadata' as a request body format.
   If you need TypeSpec support for this then please open an issue in the TypeSpec repo to request it.

File: specification/batch/Batch/models.tsp [hunk 1]
```diff
@@ -0,0 +1,5034 @@
...
+  @visibility("read")
+  publicData?: string;
+
+  @doc("This property is set only if the Certificate is in the DeleteFailed state.")
+  @visibility("read")
+  deleteCertificateError?: DeleteCertificateError;
+
+  @doc("The base64-encoded contents of the Certificate. The maximum size is 10KB.")
+  @visibility("create")
+  data: string;
```
Comment (mikekistler) [inline specification/batch/Batch/models.tsp:998]: I think properties that are base64-encoded should be defined as 'bytes'.

File: specification/batch/Batch/models.tsp [hunk 2]
```diff
@@ -0,0 +1,5034 @@
...
+  @visibility("read")
+  previousState?: CertificateState;
+
+  @doc("This property is not set if the Certificate is in its initial Active state.")
+  @visibility("read")
+  previousStateTransitionTime?: utcDateTime;
+
+  @doc("The public part of the Certificate as a base-64 encoded .cer file.")
+  @visibility("read")
+  publicData?: string;
```
Comment (mikekistler) [inline specification/batch/Batch/models.tsp:990]: Here too I think this should be `bytes`

**Review by mikekistler — COMMENTED** [review]
I happened to notice a couple properties that should be defined as `bytes` rather than `string`.

File: specification/batch/Batch/main.tsp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+@service({
+  title: "BatchServiceClient",
+})
+
+
+
+@doc("A client for issuing REST requests to the Azure Batch service.")
+@useAuth(AADToken)
+@versioned(BatchService.Versions)
```
Comment (lmazuel) [inline specification/batch/Batch/main.tsp:16]: You need a server:
```
@server(
  "{endpoint}/contentsafety",
  "Analyze harmful content",
  {
    @doc("""
Supported Cognitive Services endpoints (protocol and hostname, for example:
https://<resource-name>.cognitiveservices.azure.com).
""")
    endpoint: string,
  }
)
```

File: specification/batch/Batch/routes.tsp [hunk 4]
```diff
@@ -0,0 +1,2755 @@
...
+
+  @doc("Whether the server should return the client-request-id in the response.")
+  @header
+  @extension(
+    "x-ms-parameter-grouping",
+    {
+      postfix: "Options",
+    }
+  )
+  `return-client-request-id`?: boolean = false;
```
Comment (tadelesh) [inline specification/batch/Batch/routes.tsp:2737]: My concern: is it possible for service to default add `client-request-id` if it is set in request?

File: specification/batch/Batch/tspconfig.yaml.rej [hunk 5]
```diff
@@ -0,0 +1,9 @@
+diff a/specification/batch/Batch/tspconfig.yaml b/specification/batch/Batch/tspconfig.yaml	(rejected hunks)
+@@ -3,6 +3,5 @@
+ ]
+ options:
+   "@azure-tools/typespec-autorest":
+-    "basic-setup-py": true
+     "emitter-output-dir": "{project-root}/../data-plane/Microsoft.Batch/stable/2023-04-01.17.0/"
+     "output-file": "BatchService.json"
+\ No newline at end of file
```
Comment (msyyc) [inline specification/batch/Batch/tspconfig.yaml.rej:9]: What is the usage of this file?

File: specification/batch/Batch/routes.tsp [hunk 6]
```diff
@@ -0,0 +1,2755 @@
...
+    BatchError
+  >;
+
+  @doc("Gets basic properties of a Pool.")
+  @example("./examples/PoolExists.json", "Check Pool Exists")
+  @route("/pools/{poolId}")
+  @head
+  Exists is RpcOperation<
+    BatchClientRequestHeaders & BatchPoolHeaders,
+    PoolDoesntExistResponseHeaders | PoolDoesExistResponseHeaders,
```
Comment (msyyc) [inline specification/batch/Batch/routes.tsp:246]: I meet the following error with latest tsp compilor (0.45.2):
image
It seems illegal when anonymous model is assignable to type 'Model'

File: specification/batch/Batch/routes.tsp [hunk 7]
```diff
@@ -0,0 +1,2755 @@
...
+  @route("/jobschedules/{jobScheduleId}")
+  @head
+  Exists is RpcOperation<
+    BatchClientRequestHeaders &
+      BatchMatchHeaders & {
+        @doc("The ID of the Job Schedule which you want to check.")
+        @path
+        jobScheduleId: string;
+      },
+    (BatchResponseHeaders & {
```
Comment (msyyc) [inline specification/batch/Batch/routes.tsp:1354]: same with https://github.com/Azure/azure-rest-api-specs/pull/23291/files#r1251414153

File: specification/batch/Batch/routes.tsp (hunk 6 above)
Comment (msyyc) [reply specification/batch/Batch/routes.tsp:246]: comment from my colleague:
image

File: specification/batch/Batch/routes.tsp (hunk 4 above)
Comment (MaryGao) [reply specification/batch/Batch/routes.tsp:2737]: Why do we need this header?

File: specification/batch/Batch/routes.tsp [hunk 8]
```diff
@@ -0,0 +1,2755 @@
...
+#suppress "@azure-tools/typespec-azure-core/casing-style" "The names of Model types must use camelCase"
+@doc("Common header values for pool requests")
+model BatchMatchHeaders extends BatchModifiedSinceHeaders {
+  @doc("""
+An ETag value associated with the version of the resource known to the client.
+The operation will be performed only if the resource's current ETag on the
+service exactly matches the value specified by the client.
+""")
+  @header
+  `If-Match`?: string;
```
Comment (pshao25) [inline specification/batch/Batch/routes.tsp:2539]: Just a reminder: if 304/412 is a valid response code, you should explicitly define it as response. We don't magically add those codes as a valid response when you define etag header. Put it another way, we only take the status codes you defined in your operations as valid response codes.

**Pull request closed without merging**